apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl

```diff
@@ -1,277 +1,253 @@
-%% see rocksdb:read_options()
--type db_read_options() :: proplists:proplist().
+-type shard_id() :: emqx_ds_replication_layer:shard_id().
 
 -type cf_refs() :: [{string(), rocksdb:cf_handle()}].
 
-%% Message storage generation
-%% Keep in mind that instances of this type are persisted in long-term storage.
--type generation() :: #{
-    %% Module that handles data for the generation
-    module := module(),
-    %% Module-specific data defined at generation creation time
-    data := term(),
-    %% When should this generation become active?
-    %% This generation should only contain messages timestamped no earlier than that.
-    %% The very first generation will have `since` equal 0.
-    since := emqx_ds:time()
-}.
+-type gen_id() :: 0..16#ffff.
 
--record(s, {
-    shard :: emqx_ds:shard(),
-    db :: rocksdb:db_handle(),
-    cf_iterator :: rocksdb:cf_handle(),
-    cf_generations :: cf_refs()
+%% Note: this record might be stored permanently on a remote node.
+-record(stream, {
+    generation :: gen_id(),
+    enc :: _EncapsultatedData,
+    misc = #{} :: map()
 }).
 
+-opaque stream() :: #stream{}.
+
+%% Note: this record might be stored permanently on a remote node.
 -record(it, {
-    shard :: emqx_ds:shard(),
-    gen :: gen_id(),
-    replay :: emqx_ds:replay(),
-    module :: module(),
-    data :: term()
+    generation :: gen_id(),
+    enc :: _EncapsultatedData,
+    misc = #{} :: map()
 }).
 
--type gen_id() :: 0..16#ffff.
-
--opaque state() :: #s{}.
 -opaque iterator() :: #it{}.
 
-%% Contents of the default column family:
-%%
-%% [{<<"genNN">>, #generation{}}, ...,
-%%  {<<"current">>, GenID}]
+%%%% Generation:
 
--define(DEFAULT_CF, "default").
--define(DEFAULT_CF_OPTS, []).
+-type generation(Data) :: #{
+    %% Module that handles data for the generation:
+    module := module(),
+    %% Module-specific data defined at generation creation time:
+    data := Data,
+    %% When should this generation become active?
+    %% This generation should only contain messages timestamped no earlier than that.
+    %% The very first generation will have `since` equal 0.
+    since := emqx_ds:time(),
+    until := emqx_ds:time() | undefined
+}.
 
--define(ITERATOR_CF, "$iterators").
+%% Schema for a generation. Persistent term.
+-type generation_schema() :: generation(term()).
 
-%% TODO
-%% 1. CuckooTable might be of use here / `OptimizeForPointLookup(...)`.
-%% 2. Supposedly might be compressed _very_ effectively.
-%% 3. `inplace_update_support`?
--define(ITERATOR_CF_OPTS, []).
+%% Runtime view of generation:
+-type generation() :: generation(term()).
 
--define(REF(Keyspace, ShardId), {via, gproc, {n, l, {?MODULE, Keyspace, ShardId}}}).
+%%%% Shard:
+
+-type shard(GenData) :: #{
+    %% ID of the current generation (where the new data is written:)
+    current_generation := gen_id(),
+    %% This data is used to create new generation:
+    prototype := {module(), term()},
+    %% Generations:
+    {generation, gen_id()} => GenData
```

Comment:
Might be worth to also keep `{generation, current} => GenData` here as a shortcut?

Reply:
Maybe in the next PR.